### PR TITLE
Port to Python>=3.4 and make more Pythonic/maintainable

### PR DESCRIPTION
**Pull request by tmose1106**
## Commit Message

This commit includes all of the changes I have made to the Metrics tool
over the past few months. Everything was modified in some way to make
the overall project more maintainable and conform to standard Python 3
conventions. With the old package being stuck with Python 2, I have
updated it so that it now functions under Python 3.4 and greater. All in
all, I believe the quality of the code has increased greatly and it is
much more manageable. Finally, the original tests are back and even more
have been added which all work under recent versions of Pytest.

## Note to Maintainer
Hello, over the past few months I have worked on improving Metrics from
time to time. As stated in the commit message, I have ported the project to
work with Python>=3.4 ONLY. By doing so, the code is now much more
modern-looking Python. I have also added several tests, which can be run
using [Pytest](https://docs.pytest.org/en/latest/) after installing the package
along with its two dependencies (listed in `requirements.txt`). I would
appreciate it if you could take some time to look over my changes and if you
have any requests, I am open to ideas. Thank you for your time in advance.

### CODE REVIEW

**Comment by markfink** [comment]
usually pull requests are much smaller than so this might take a few days to review but browsing thought the code it looks like it is in great shape.
I want to run the tests and see for myself. I think you know that I started that back in 2009/2010. Python 3 was not really a thing back then. Python 2.7 today is still a must have (at least in the enterprise domain) but this should be possible. I do not think I would like the click thing so I probably like to go with docopts or simple argparse.

**Comment by tmose1106** [comment]
## Python 2 Compatibility

If you were interested in making this code Python 2 compatible, you could
probably do so fairly easy. Between using the ``futures`` in 2.7 and possibly
using the ``six`` library. There are a few spots I can think of that are likely
to cause issues for Python 2:

* The change in syntax for error statements may cause an issue.
* A few times I use _dict.items()_ whereas in Python 2 it is _dict.iteritems()_.

But other than those, everything should be mostly working.

## Click For Argument Parsing

On this topic, I will not say you are wrong. I see the merit in using
``argparse`` to be more standard. But let me explain why I think Click is a
good dependency.

At first the wrapper syntax heavily used by the ``Click`` module can quite
confusing and messy looking. But I think it works well enough, and I tend to
use it in all of my command line tools. But I do not think that argument
parsing is its biggest selling point. Throughout the *main.py* script, you may
notice that there are no _print()_ statements, and I have substituted them with
_click.echo()_ statements. From what I understand, the _click.echo()_ statement
acts exactly the same across Python 2 and Python 3 and it handles any character
encoding, unicode, bytes, that is necessary to print the correct text to the
command line. Click also has _click.style()_ which handles color output across
platforms, which I think is an under-utilized feature of the command line. So
as discussed above, if you want compatibility all across the program for
Python 2, I would highly suggest using ``Click``.

**Comment by markfink** [comment]
I was thinking about this a bit longer and since you are the 2nd person in almost 10 years who takes an interest in this library I hope you do not mind me asking why you are doing this?
Do you by any chance have a working jupyter notebook & bqplot which demonstrates how to put the metrics library to use? I think updating it to python3 is important and all but demonstrating how to use it with current visualization technology is even more important.
Back in 2009 I used a different stack since there was no jupyter notebook but the paper still might be useful to convey the idea:
[mark_fink_visualizing_software_quality_paper.pdf](https://github.com/markfink/metrics/files/1033657/mark_fink_visualizing_software_quality_paper.pdf)

**Comment by tmose1106** [comment]
The reason I came across this tool is actually kinda interesting. I am a pretty 
big user of [Gitlab](https://about.gitlab.com/) over GitHub since Gitlab is 
actually an open source project and I am very much for software freedom. But 
one of the things that bugs me is that Gitlab doesn't have that colorful 
breakdown of languages bar like GitHub does. So I wanted to find a tool which 
could tell me what programming languages were used in basically any open source 
project.

Since I am well acquainted with Python, I went on GitHub searching for a SLOC 
tool which I could perhaps modify. I also went on PyPi and I came across your 
Metrics project which seemed very much in line with what I wanted to do, so I 
began hacking on it. I also am interested in porting/cloning some larger 
programs into Python 3, and Metrics seemed like a good project to start 
practicing on.

Despite using Python all the time, I have never used a Jupyter notebook. Though 
I am aware of what they are, and after reading the first third of your paper 
(I'll probably read the rest later), it makes a lot more sense why you would 
make Metrics.

So, in summary, I am pretty much only interested in Metrics for the sake of 
curiosity, and I would like to use it myself for looking over source code repos 
via the command line. If you feel that our ideas of this project's intended use 
differ, I will be fine with just forking off and changing the name. But I am 
still very much interested in the PyPi page since I feel "Metrics" is a very 
appropriate name for the project, and PyPi is the standard way to download 
Python modules and tools.

Of course I don't mind contributing back here via GitHub, but not necessarily
if you don't agree with the direction in which I want to take the project. I
still have some ideas of other things I can add to make the project better for
my use case.

**Comment by markfink** [comment]
I have never thought about the metrics library in that way but I get your point. Why don't you contribute the language breakdown to Gitlab? To me this would make a lot of sense since it is open source and free.

**Comment by tmose1106** [comment]
As far as I know, there is already a language breakdown in the form of a pie chart under Repository -> Charts. So the data is there, it just takes a bit of looking. Whereas on GitHub, that bar can instantly tell me everything I need to know given I have memorized what quite a few of the colors mean. So if I went through the process, I could probably add that bar without worrying about Metrics at all. Or perhaps I could do as you suggest and supply Gitlab with even more information about projects. 

I was previously aware that they had this feature, but I was sick of worrying about it, and wanted something that always worked, no matter the site it is on or how I get the code. And now I am able to view the information I want and more. 

I see that I should at least raise an issue with Gitlab so that the user experience could be improved for everyone using Gitlab. But in the end, I have had a good time working on Metrics, and I learned some new things like how to use subclassing a bit and also some differences between/techniques for porting Python 2 and 3.

**Comment by markfink** [comment]
@tmose1106 hi , I did not hear back from you for some time. Hope you are doing well.

I hope where able to push your ideas and code into Gitlab like we discussed earlier. 

Regarding metrics I just wanted to let you know that I am about to release some dependency updates and some fixes. Besides that the only thing I wanted to change was to move from nose to pytest. Hopefully I will manage to release a new version of metrics containing these changes within the next few days.

What do you think? Can we close this?

**Comment by tmose1106** [comment]
If you want to apply your own changes, then I will be forking off into a new project and that is that. So yes, we can close this pull request. 

As for Gitlab, I have yet to discuss anything with them, but it is certainly still on my to-do.

Thanks for your time.
